[obsolete] Logic: algebraic normal form

konstantintogoi opened the pull request:
Added new features for converting boolean expression to
Algebraic Normal Form.

New functions:
- to_anf;
- is_anf;
- ditribute_xor_over_and;
- _convert_to_varsANF;
- anf_coeffs;
- ANFform;
- minterm;
- maxterm;
- monomial;

New classmethods:
- BooleanFunction.to_anf;
- BooleanFunction._to_anf;
- And.to_anf;
- Or.to_anf;
- Not.to_anf;
- Implies.to_anf;
- Equivalent.to_anf;

asmeurer commented on sympy/logic/boolalg.py:
```diff
@@ -1637,6 +1808,84 @@ def POSform(variables, minterms, dontcares=None):
+    return Xor(*[_convert_to_varsANF(x, variables) for x in terms])
+
+
+def coeffs(truthvalues):
```
"Coeffs" can mean a lot of things in SymPy. Is there a more specific name we can use for this? Maybe `zhegalkin_coeffs`.

KoStiG replied:
The name <code>zhegalkin_coeffs</code> is too long. <code>anf_coeffs</code> is prettier and shows that it's related to <code>ANF</code> function. Both of these functions, in fact, returns the same thing, the boolean expression in algebraic normal form. Just in different representations.

asmeurer replied:
`anf_coeffs` sounds fine to me.

asmeurer commented on sympy/logic/boolalg.py:
```diff
@@ -1637,6 +1808,84 @@ def POSform(variables, minterms, dontcares=None):
+    >>> from sympy.logic.boolalg import coeffs
+    >>> truthvalues = [0, 1, 1, 0, 0, 1, 0, 1]
+    >>> coeffs(truthvalues)
+    [0, 1, 1, 0, 0, 0, 1, 0]
```
Maybe show how this example corresponds to a polynomial.

KoStiG replied:
In order to represent the Boolean expression in Algebraic Normal Form (ANF) it is enough to know the list of indexes of monomials with non zero coefficients.

For example, the list <code>[1, 2, 6]</code> can be considered as the Boolean expression with three variables in ANF.  In this case, ANF contains only 1-st, 2-nd and 6-th monomials. We can rewrite the list like this <code>[0, 1, 1, 0, 0, 0, 1, 0]</code>.

Monomial can be deduced from its index. We need to convert the index of monomial from decimal to binary. For example, <code>[1, 2, 6] -> [[0, 0, 1], [0, 1, 0], [1, 1, 0]] </code>. If our variables <code>[a, b, c]</code>, then monomials <code>a<sup>0</sup>b<sup>0</sup>c<sup>1</sup> = c </code>; <code>a<sup>0</sup>b<sup>1</sup>c<sup>0</sup> = b </code>; <code>a<sup>1</sup>b<sup>1</sup>c<sup>0</sup> = ab </code>. ANF in this example: <code>Xor(b, c, And(a, b))</code>.

Converting truth values to ANF is similar to converting minterms to SOP form, or converting maxterms to POS form.
In order to convert minterms to SOP form, we only need to know the indexes of the minterms. In order to convert maxterms to POS form, we only need to know the indexes of the maxterms.

Converting minterms to SOP form:
<code> [1, 2, 4, 7] -> [[0, 0, 1], [0, 1, 0], [1, 0, 0], [1, 1, 1]] -> (~a & ~b & c) | (~a & b & ~c) | (a & ~b & ~c) | (a & b & c) </code>

asmeurer replied:
I would put this whole comment in the docstring, or somewhere else in the docs.

asmeurer commented on sympy/logic/boolalg.py:
```diff
@@ -1637,6 +1808,84 @@ def POSform(variables, minterms, dontcares=None):
+
+    Return True, False, logical And function (i.e., the
+    "Zhegalkin monomial") or logical Xor function (i.e.,
+    the "Zhegalkin polynomial").
```
Maybe point out that this is because when True and False are represented by 1 and 0, respectively, then And is multiplication and Xor is addition.

asmeurer commented on sympy/logic/boolalg.py:
```diff
@@ -994,6 +1061,40 @@ def _distribute(info):
+    conjunction of variables. No negations are permitted.
+
+    If ``complete`` is false, arguments of the boolean
+    expression considered as variables.
```
I'm confused what this means. Does it mean that if `complete=False`, then only the top-level expression is converted to ANF? If so, a more standard SymPy name for this is `deep=False`.

KoStiG replied:
Yes, exactly. I didn't know how to name it properly.

I used the conversions:
- Or(x<sub>1</sub> , x<sub>2</sub>) = Xor(x<sub>1</sub> , x<sub>2</sub>, And(x<sub>1</sub> , x<sub>2</sub>));
- Or(x<sub>1</sub> , x<sub>2</sub>, ... , x<sub>n</sub>) = Xor(x<sub>1</sub>, Or(x<sub>2</sub>, ... , x<sub>n</sub>), And(x<sub>1</sub>, Or(x<sub>2</sub>, ... , x<sub>n</sub>)));
- Not(x) = Xor(x, True);
- Implies(a, b) = Xor(a, True, And(a, b));
- Equivalent(x<sub>1</sub> , x<sub>2</sub>) = Or(And(x<sub>1</sub> , x<sub>2</sub>), And(~x<sub>1</sub> , ~x<sub>2</sub>)) = Xor(x<sub>1</sub> , x<sub>2</sub>, True);
- Equivalent(x<sub>1</sub> , x<sub>2</sub>, ... , x<sub>n</sub>) = Or(And(x<sub>1</sub> , x<sub>2</sub>, ... , x<sub>n</sub>), And(~x<sub>1</sub> , ~x<sub>2</sub>, ... , ~x<sub>n</sub>)) = Xor(x<sub>1</sub> , x<sub>2</sub>, ... , x<sub>n</sub>, True, ....)

asmeurer commented on sympy/logic/boolalg.py:
```diff
@@ -624,8 +656,15 @@ def __new__(cls, *args, **kwargs):
-            argset.remove(True)
-            return Not(Xor(*argset))
+            frame = stack()[1][0]
+            frame_info = getframeinfo(frame)
```
Woah! I don't like this. What are you trying to achieve here?

KoStiG replied:
By default, if True is in argset, Xor's constructor removes True and returns Not. I use functions stack() and getframeinfo() to identify the name of the function which calls Xor's contructors. If function's name in ['_to_anf', '_distribute', 'ANF'], Xor's constructor returns Xor with True in argset.

asmeurer commented:
Thanks for the pull request. I only have those comments.

konstantintogoi committed:
Logic: algebraic normal form

Added new features for converting boolean expression to
Algebraic Normal Form.

New functions:
- to_anf;
- is_anf;
- ditribute_xor_over_and;
- _convert_to_varsANF;
- anf_coeffs;
- ANF;
- minterm;
- maxterm;
- monomial;

New classmethods:
- BooleanFunction.to_anf;
- BooleanFunction._to_anf;
- And.to_anf;
- Or.to_anf;
- Not.to_anf;
- Implies.to_anf;
- Equivalent.to_anf;

konstantintogoi committed:
edits of boolalg and test_boolalg

- modified Xor's __new__ method in boolalg (added usage of
    keyword argument ``remove_true``);
- modified test_to_anf in test_boolalg (used keyword argument
    ``remove_true`` in Xor function);

KoStiG commented:
@asmeurer , please review.

asmeurer commented on sympy/logic/boolalg.py:
```diff
@@ -416,6 +438,12 @@ def as_set(self):
                                       " expressions")
 
+    def to_anf(self, deep=True):
+        a = self.args[0]
```
Will the final result depend on what args[0] happens to be?

asmeurer commented on sympy/logic/boolalg.py:
```diff
@@ -1637,6 +1807,188 @@ def POSform(variables, minterms, dontcares=None):
+    >>> from sympy.logic.boolalg import ANF
+    >>> ANF(['x'], [1, 0])
+    Xor(x, True)
+    >>> ANF(['x', 'y'], [0, 1, 1, 1])
```
Don't encourage the use of strings. Use `x, y = symbols('x y')`.

asmeurer commented on sympy/logic/boolalg.py:
```diff
@@ -1637,6 +1807,188 @@ def POSform(variables, minterms, dontcares=None):
+    >>> coeffs = anf_coeffs(truthvalues)
+    >>> coeffs
+    [0, 1, 1, 0, 0, 0, 1, 0]
+    >>> polynomial = Xor(*[monomial(k, ['a','b','c'])
```
Same here.

asmeurer commented on sympy/logic/boolalg.py:
```diff
@@ -1637,6 +1807,188 @@ def POSform(variables, minterms, dontcares=None):
+    >>> from sympy.abc import x, y, z
+    >>> minterm([1, 0, 1], [x, y, z])
+    And(Not(y), x, z)
+    >>> minterm(6, ['x', 'y', 'z'])
```
And here. Just use `minterm(6, [x, y, z])`.

asmeurer commented on sympy/logic/boolalg.py:
```diff
@@ -1637,6 +1807,188 @@ def POSform(variables, minterms, dontcares=None):
+    >>> from sympy.abc import x, y, z
+    >>> maxterm([1, 0, 1], [x, y, z])
+    Or(Not(x), Not(z), y)
+    >>> maxterm(6, ['x', 'y', 'z'])
```
And here.

asmeurer commented on sympy/logic/boolalg.py:
```diff
@@ -1637,6 +1807,188 @@ def POSform(variables, minterms, dontcares=None):
+    >>> from sympy.abc import x, y, z
+    >>> monomial([1, 0, 1], [x, y, z])
+    And(x, z)
+    >>> monomial(6, ['x', 'y', 'z'])
```
And here.

asmeurer commented on sympy/logic/boolalg.py:
```diff
@@ -1637,6 +1807,188 @@ def POSform(variables, minterms, dontcares=None):
+
+def monomial(k, variables):
+    """
+    Return the k-th monomial.
```
Please explain what this means.

KoStiG replied:
Each boolean function can be uniquely represented by a _Zhegalkin Polynomial_ (Algebraic Normal Form). 
The Zhegalkin Polynomial of the boolean function with _n_ variables can contain up to _2<sup>n</sup>_ monomials.
We can enumerate all the monomials. Each monomial is fully specified by the presence or absence of each variable.
For example, boolean function with four variables _(a, b, c, d)_ can contain up to 2<sup>4</sup>=16 monomials. The 13-th monomial is the product _abd_, because 13 in binary is _1, 1, 0, 1_.

asmeurer commented on sympy/logic/boolalg.py:
```diff
@@ -1637,6 +1807,188 @@ def POSform(variables, minterms, dontcares=None):
+
+def maxterm(k, variables):
+    """
+    Return the k-th maxterm.
```
Please explain what this means.

KoStiG replied:
Function <code>maxterm</code> returns a logical expression of _n_ variables that employs only the _complement_ operator and the _disjunction_ operator.

There are 2<sup>n</sup> maxterms of n variables, since a variable in the maxterm expression can be in either its direct or its complemented form—two choices per n variables.

Complementation pattern of the _k-th_ maxterm can be deduced from the base-2 digits in the integer _k_.

Reference - http://en.wikipedia.org/wiki/Canonical_normal_form#Indexing_maxterms

Function <code>maxterm</code> is similar to function <code>BooleanMaxterms</code> in Wolfram Mathematica http://reference.wolfram.com/language/ref/BooleanMaxterms.html

asmeurer commented on sympy/logic/boolalg.py:
```diff
@@ -1637,6 +1807,188 @@ def POSform(variables, minterms, dontcares=None):
+
+def minterm(k, variables):
+    """
+    Return the k-th minterm.
```
Please explain what this means.

KoStiG replied:
Function <code>minterm</code> returns a logical expression of _n_ variables that employs only the _complement_ operator and the _conjunction_ operator.

There are 2<sup>n</sup> minterms of n variables, since a variable in the minterm expression can be in either its direct or its complemented form—two choices per n variables.

Complementation pattern of the _k-th_ minterm can be deduced from the base-2 digits in the integer _k_.

Reference - http://en.wikipedia.org/wiki/Canonical_normal_form#Indexing_minterms

Function <code>minterm</code> is similar to function <code>BooleanMinterms</code> in Wolfram Mathematica http://reference.wolfram.com/language/ref/BooleanMinterms.html

asmeurer commented on sympy/logic/tests/test_boolalg.py:
```diff
@@ -708,3 +710,65 @@ def test_truth_table():
+
+def test_maxterm():
+    x, y = symbols('x,y')
+    assert maxterm(2, ['x', 'y']) == Or(Not(x), y)
```
Don't use strings in the tests either. Strings as input is something that should only work incidentally. See https://github.com/sympy/sympy/wiki/Idioms-and-Antipatterns#strings-as-input.

asmeurer commented:
I have only a few comments and questions. Mostly documentation related.

asmeurer commented on sympy/logic/boolalg.py:
```diff
@@ -1637,6 +1807,188 @@ def POSform(variables, minterms, dontcares=None):
     return And(*[_convert_to_varsPOS(x, variables) for x in essential])
 
 
+def ANF(variables, truthvalues):
```
Is this the analogue of `POSform`? Maybe it should be called `ANFform`?

asmeurer commented:
I see now that you were just following what was already there for the strings. I opened https://github.com/sympy/sympy/pull/9117 to remove existing instances of strings as input.

konstantintogoi committed:
Logic: docstring updated, ANF -> ANFform

- ANF was renamed ANFform;
- docstring updated in ANFform, anf_coeffs,
minterm, maxterm, monomial;

konstantintogoi committed:
Logic: Or.to_anf() updated, added spaces in test_boolalg

- conversion of Or function to ANF was changed;

sleveque commented:
+1 for this.

Should "anf" be an accepted form in [simplify_logic](http://docs.sympy.org/latest/modules/logic.html#sympy.logic.boolalg.simplify_logic)?
